File: datastructures/java/assignments/assignment1_CommonElements/CommonElements.java
import java.util.Scanner;

public class CommonElements {
    public static void main(String[] args) {
        int n, m, k = 0;
        // int[] arr1 = { 1, 5, 6, 6, 9, 9, 9, 11, 11, 21 };
        // int[] arr2 = { 6, 6, 9, 11, 21, 21, 21 };
        System.out.println("Array 1:");
        int[] arr1 = makeArray();
        System.out.println("Array 2:");
        int[] arr2 = makeArray();
        int[] array = new int[15];
        n = removeDuplicatesOptimized(arr1, arr1.length);
        m = removeDuplicatesOptimized(arr2, arr2.length);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (arr1[i] == arr2[j]) {
                    array[k] = arr1[i];
                    k++;
                }
            }
        }
        System.out.println("elements of the array are: ");
        for (int i = 0; i < k; i++)
            System.out.print(array[i] + " ");
    }

    public static int removeDuplicatesOptimized(int[] arr, int n) {
        // if array has a length of 0 or 1 then it has no duplicates --test case
        if (n == 0 || n == 1) {
            return n;
        }

        int uniques = 0;

        // if i!=i+1 then it is unique so place i at the next unique index of the array
        for (int i = 0; i < n - 1; i++) {
            if (arr[i] != arr[i + 1]) {
                arr[uniques++] = arr[i];
            }
        }

        // add the last unique number
        arr[uniques++] = arr[n - 1];

        return uniques;
    }

    public static int[] makeArray() {
        Scanner scanner = new Scanner(System.in);

        // user decides how long they want the array
        System.out.print("How long would you like to make your array? ");
        int length = scanner.nextInt();

        // if user enters a non positive number, prompt them to enter a positive
        while (length <= 0) {
            System.out.println("Please enter a positive number. ");
            length = scanner.nextInt();
        }

        // create a new array at the length specified by the user
        int[] arr = new int[length];

        // have the user enter all the elements for the array
        for (int i = 0; i < length; i++) {
            System.out.println("Enter element " + i + " or your array. ");
            arr[i] = scanner.nextInt();
        }
        return arr;
    }
}

File: datastructures/python/assignments/assignment1_CommonElements/CommonElements.py
def make_array() -> list[int]:
    # user decides how long they want the array
    length = int(input("How long would you like to make your array? "))

    # if user enters a non positive number, prompt them to enter a positive
    while length <= 0:
        length = int(input("Please enter a positive number. "))

    arr = []
    # have the user enter all the elements for the array
    for i in range(length):
        arr.append(int(input(f"Enter element {i+1} for your array. ")))

    return arr


def remove_duplicates_optimized(arr: list[int], list_length: int) -> int:
    # if array has a length of 0 or 1 then it is sorted already
    if list_length == 0 or list_length == 1:
        return list_length
    uniques = 0

    # if i!=i+1 then it is unique so place i
    # at the next unique index of the array
    for i in range(list_length - 1):
        if arr[i] != arr[i + 1]:
            arr[uniques] = arr[i]
            uniques += 1

    # add last unique number and increment counter
    arr[uniques] = arr[-1]
    uniques += 1

    return uniques


if __name__ == "__main__":
    print("Array 1:")
    arr1 = make_array()
    print("Array 2:")
    arr2 = make_array()
    common_array = []

    arr1_uniques = remove_duplicates_optimized(arr1, len(arr1))
    arr2_uniques = remove_duplicates_optimized(arr2, len(arr2))

    for i in arr1[:arr1_uniques]:
        for j in arr2[:arr2_uniques]:
            if i == j:
                common_array.append(i)

    print(f"The common elements are {common_array}")
